### DISCUSSION TITLE
ScriptReference/SettingsProvider

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#247] [Notes]
# ScriptReference/SettingsProvider

https://docs.unity3d.com/ScriptReference/SettingsProvider.html

**Comment by vertxxyz** [comment]
## Creating project settings assets
### Setup
To create a settings file in the Project Settings directory you can instead inherit your settings ScritableObject from [`ScriptableSingleton`](https://docs.unity3d.com/2020.1/Documentation/ScriptReference/ScriptableSingleton_1.html) with a [`FilePath`](https://docs.unity3d.com/2020.1/Documentation/ScriptReference/FilePathAttribute.html) of `ProjectSettings/Example.asset`.

### PropertyField issues
The following has been [fixed](https://issuetracker.unity3d.com/issues/setting-scriptablesingletons-hideflags-on-its-awake-or-onenable-is-ignored-when-the-serialized-file-does-not-exist) as of 2022.3.7f1, 2023.1.7f1, and 2023.2.0b3.
> The fields will not be editable, so you have to fix that, in [`OnActivate`](https://docs.unity3d.com/ScriptReference/SettingsProvider.OnActivate.html) you can run some code like:
> ```csharp
> ExampleSettings settings = ExampleSettings.instance;
> settings.hideFlags = HideFlags.DontSave | HideFlags.HideInHierarchy;
> settings.Save();
> _serializedObject = new SerializedObject(settings);
> ```
> Changing its hideFlags so it's editable. And calling `Save` to force the asset to exist.
> Save is a custom function that calls through to the protected [`Save`](https://docs.unity3d.com/2020.1/Documentation/ScriptReference/ScriptableSingleton_1.Save.html):
> ```csharp
> public void Save() => Save(true);
> ```

### Usage
Use `_serializedObject ` as you would in any other Editor, but calling `Save` after `ApplyModifiedProperties`.

You should now have a `.asset` file in your ProjectSettings directory and a functioning SettingsProvider to display it.
